### ISSUE TITLE
CUDA sum incorrect on V100

### ISSUE THREAD

**Issue by charleskawczynski** [CliMA/ClimaCore.jl#1240]
## Describe the bug
The CUDA `sum` is producing different results between CPU and GPU for certain spaces. See [this build](https://buildkite.com/clima/climacore-ci/builds/2052#01882b5c-45b4-4528-a55e-88200c868acb), which is the result of widening the test spaces in [this PR](https://github.com/CliMA/ClimaCore.jl/pull/1239). For now, I'll check the array type and note these as broken tests.

cc @sriharshakandala, @simonbyrne

**Comment by simonbyrne** [comment]
@sriharshakandala can you please look into this ASAP?

**Comment by sriharshakandala** [comment]
These tests seem to be running correctly on the GPU cluster (single GPU). Please see below:
```
[skandala@hpc-26-18 test]$ git status
On branch ck/test_utils
Your branch is up to date with 'origin/ck/test_utils'.

nothing to commit, working tree clean
[skandala@hpc-26-18 test]$ srun julia --project Operators/finitedifference/column.jl 
Scalar Field FiniteDifferenceSpaces: Error During Test at /central/home/skandala/ClimaCore.jl/test/Operators/finitedifference/column.jl:36
 Unexpected Pass
 Expression: sum(fones) ≈ pi
 Got correct result, please change to @test if no longer broken.

Scalar Field FiniteDifferenceSpaces: Error During Test at /central/home/skandala/ClimaCore.jl/test/Operators/finitedifference/column.jl:47
 Unexpected Pass
 Expression: ≈(sum(ffield), FT(2.0), atol = 0.01)
 Got correct result, please change to @test if no longer broken.

Scalar Field FiniteDifferenceSpaces: Error During Test at /central/home/skandala/ClimaCore.jl/test/Operators/finitedifference/column.jl:36
 Unexpected Pass
 Expression: sum(fones) ≈ pi
 Got correct result, please change to @test if no longer broken.

Scalar Field FiniteDifferenceSpaces: Error During Test at /central/home/skandala/ClimaCore.jl/test/Operators/finitedifference/column.jl:47
 Unexpected Pass
 Expression: ≈(sum(ffield), FT(2.0), atol = 0.01)
 Got correct result, please change to @test if no longer broken.

Test Summary:                       | Pass  Error  Total   Time
Scalar Field FiniteDifferenceSpaces |   18      4     22  59.3s
ERROR: LoadError: Some tests did not pass: 18 passed, 0 failed, 4 errored, 0 broken.
in expression starting at /central/home/skandala/ClimaCore.jl/test/Operators/finitedifference/column.jl:12
srun: error: hpc-26-18: task 0: Exited with exit code 1
[skandala@hpc-26-18 test]$ 
```

**Comment by sriharshakandala** [comment]
cc: @charleskawczynski @simonbyrne

**Comment by charleskawczynski** [comment]
Can you try the branch where the failure occurred (https://github.com/CliMA/ClimaCore.jl/pull/1239/commits/5223f0f759f718a8e2d311038c15edb98b293d1b)? The unexpected pass means that you ran onhttps://github.com/CliMA/ClimaCore.jl/pull/1239/commits/2d32cb21d78da130872d7a6ba50f5e5522da3fbf

**Comment by sriharshakandala** [comment]
Please see below:
```
[skandala@hpc-26-18 test]$ git status
HEAD detached at 5223f0f7
nothing to commit, working tree clean
[skandala@hpc-26-18 test]$ srun julia --project Operators/finitedifference/column.jl 
Test Summary:                       | Pass  Total   Time
Scalar Field FiniteDifferenceSpaces |   22     22  59.1s
Test Summary:                                  | Pass  Total  Time
Scalar Field FiniteDifferenceSpaces - periodic |   20     20  9.5s
┌ Warning: Bounds check on BoundsError ∂.(w .* I.(θ)) not verified.
└ @ Main /central/home/skandala/ClimaCore.jl/test/Operators/finitedifference/column.jl:208
┌ Warning: Bounds check on BoundsError ∂.(w .* I.(θ)) not verified.
└ @ Main /central/home/skandala/ClimaCore.jl/test/Operators/finitedifference/column.jl:208
Test Summary:          | Pass  Total  Time
Test composed stencils |   14     14  5.3s
Test Summary:                          | Pass  Total  Time
Composite Field FiniteDifferenceSpaces |   14     14  0.6s
Test Summary:                                        | Pass  Total  Time
Face -> Center interpolation (uniform and stretched) |    4      4  2.2s
Test Summary:                                        | Pass  Total  Time
Center -> Face interpolation (uniform and stretched) |    4      4  1.5s
Test Summary:                                          | Pass  Total  Time
∂ Center -> Face interpolation (uniform and stretched) |   10     10  1.2s
Test Summary:                                          | Pass  Total  Time
∂ Face -> Center interpolation (uniform and stretched) |   10     10  1.0s
Test Summary:                                  | Pass  Total  Time
∂ Center -> Face and ∂ Face-> Center (uniform) |   35     35  2.1s
Test Summary:                                                                         | Pass  Total  Time
Upwind3rdOrderBiasedProductC2F + DivergenceF2C on (uniform) periodic mesh, constant w |    5      5  0.0s
Test Summary:                                                                             | Pass  Total  Time
Upwind3rdOrderBiasedProductC2F + DivergenceF2C on (uniform) periodic mesh, varying sign w |    4      4  0.0s
Test Summary:                                                                                                                                                 | Pass  Total  Time
Upwind3rdOrderBiasedProductC2F + DivergenceF2C on (uniform and stretched) non-periodic mesh, with FirstOrderOneSided + DivergenceF2C SetValue BCs, constant w |    8      8  2.2s
Test Summary:                                                                                                                                                     | Pass  Total  Time
Upwind3rdOrderBiasedProductC2F + DivergenceF2C on (uniform and stretched) non-periodic mesh, with ThirdOrderOneSided + DivergenceF2C SetValue BCs, varying sign w |    8      8  1.0s
Test Summary:                                                                                                     | Pass  Total  Time
Simple FCT: lin combination of UpwindBiasedProductC2F + Upwind3rdOrderBiasedProductC2F on (uniform) periodic mesh |    5      5  0.0s
Test Summary:                                                                                                                                                    | Pass  Total  Time
Simple FCT: lin combination of UpwindBiasedProductC2F + Upwind3rdOrderBiasedProductC2F on (uniform and stretched) non-periodic mesh, with FirstOrderOneSided BCs |    8      8  1.8s
Test Summary:                                                                                                                                                    | Pass  Total  Time
Simple FCT: lin combination of UpwindBiasedProductC2F + Upwind3rdOrderBiasedProductC2F on (uniform and stretched) non-periodic mesh, with ThirdOrderOneSided BCs |    8      8  0.5s
Test Summary:        | Pass  Total  Time
Biased interpolation |    4      4  1.5s
Test Summary:              | Pass  Total  Time
Center -> Center Advection |    4      4  0.4s
[skandala@hpc-26-18 test]$ 
```

**Comment by charleskawczynski** [comment]
Was that a GPU run?

**Comment by sriharshakandala** [comment]
> Was that a GPU run?

Yes.

**Comment by charleskawczynski** [comment]
What are the srun options you're using?

**Comment by sriharshakandala** [comment]
I am using 
```
#SBATCH --time=1:30:00
#SBATCH --ntasks=1
#SBATCH --gpus=1

export JULIA_CUDA_MEMORY_POOL=none

module purge
module load julia/1.8.5 ucx/1.13.1_cuda-11.2 cuda/11.2 openmpi/4.1.5_cuda-11.2
module load hdf5/1.12.2-ompi415 nsight-systems/2022.2.1
```
on the cluster!

**Comment by charleskawczynski** [comment]
I was unable to reproduce the error in the build of the original post, so I’ve adjusted the title accordingly.

**Comment by charleskawczynski** [comment]
Yep, here is another failure: https://buildkite.com/clima/climacore-ci/builds/2105#_

that branch did not touch src/, nor test/

**Comment by simonbyrne** [comment]
Here's another one:
https://buildkite.com/clima/climacore-ci/builds/2109#01885a83-ce89-4d6d-b546-1eafb79c91bf/141-260

**Comment by simonbyrne** [comment]
What's odd is that when one test fails, they all fail. Also, the returned value is always the same. This suggests to me it isn't a race condition in `sum`.

**Comment by simonbyrne** [comment]
It could be related to https://github.com/JuliaGPU/CUDA.jl/issues/1875?

**Comment by simonbyrne** [comment]
or perhaps https://github.com/JuliaGPU/CUDA.jl/issues/1746#issuecomment-1516825966?

**Comment by simonbyrne** [comment]
Okay, what's interesting is that all the failures linked above on a V100: this is the _opposite_ of the linked Julia issue, which only flagged it on Pascal and earlier. I wonder if using a newer CUDA would fix it.

**Comment by simonbyrne** [comment]
Okay, can confirm that the issue arises on the V100 with CUDA 11.3. It appears to go away when switching to CUDA 11.8 (CUDA 12.0 doesn't appear to be supported by the current drivers). I'll ask IMSS to install it, and compile UCX, MPI, etc.

**Comment by simonbyrne** [comment]
Fixed by #1314.
